GroupDNA_Shreya_DS22577.ipynb

# ============================================================
#                 GROUPDNA - WHATSAPP GROUP ANALYSIS
# ============================================================


Project Title:
GroupDNA - WhatsApp Group Analysis

Objective:
The objective of this project is to analyze a WhatsApp group chat
using Python fundamentals and NumPy. The project extracts useful
insights from the chat such as group activity, participant
contributions, active days and hours, word usage, response times,
silent periods, and participant personality archetypes.

Technologies Used:
- Python
- NumPy
- datetime

Key Features:
1. WhatsApp Chat Parser
2. Group Overview
3. Most Active Day and Hour
4. NumPy-based Activity Analysis
5. Top Word Frequency Analysis
6. Response Time and Silent Streak Analysis
7. Personality Archetype Detection
8. Final GroupDNA Report

Dataset:
The analysis is performed on a WhatsApp group chat export
containing messages from multiple participants.

Project Constraints:
- Only Python fundamentals and NumPy are used.
- No pandas, matplotlib, seaborn, plotly, Counter,
  defaultdict, regex, or machine learning libraries are used.
- datetime is used only where required for timestamp parsing
  and time-gap calculations.


In [ ]:
# FEATURE.1- CHAT PARSER
file_path ="/content/hostel_bois.txt"

with open(file_path, "r", encoding="utf-8") as file:
    chat = file.readlines()

print("Total lines:", len(chat))

Total lines: 3178


In [ ]:
# Checking the first 10 lines of the chat

for line in chat[:10]:
    print(line)

01/04/24, 01:12 - Messages and calls are end-to-end encrypted. No one outside of this chat, not even WhatsApp, can read or listen to them. Tap to learn more.

01/04/24, 01:13 - Priya created group "Hostel Bois 4ever"

01/04/24, 01:14 - Priya added you

01/04/24, 01:17 - Rahul: scene fix

01/04/24, 01:17 - Rahul: haan

01/04/24, 01:18 - Rahul: kya scene

01/04/24, 02:13 - Rahul: abhi free hai?

01/04/24, 02:13 - Rahul: abey

01/04/24, 02:14 - Rahul: bhai sun na

01/04/24, 07:46 - Priya: I have extra notes if anyone wants



In [ ]:
# Checking the lines that contain a Sender and message

for line in chat[:10]:
    if " - " in line and ": " in line:
        print(line.strip())

01/04/24, 01:17 - Rahul: scene fix
01/04/24, 01:17 - Rahul: haan
01/04/24, 01:18 - Rahul: kya scene
01/04/24, 02:13 - Rahul: abhi free hai?
01/04/24, 02:13 - Rahul: abey
01/04/24, 02:14 - Rahul: bhai sun na
01/04/24, 07:46 - Priya: I have extra notes if anyone wants


In [ ]:
line = "01/04/24, 01:17 - Rahul: scene fix"

parts = line.split(" - ")

date, time = parts[0].split(", ")
Sender, message = parts[1].split(": ", 1)

print("Date:", date)
print("Time:", time)
print("Sender:", Sender)
print("Message:", message)

Date: 01/04/24
Time: 01:17
Sender: Rahul
Message: scene fix


In [ ]:
messages=[]
for line in chat:
    if " - " in line and ": " in line:
        parts = line.split(" - ", 1)
        date, time = parts[0].split(", ")
        Sender, message = parts[1].split(": ", 1)
        messages.append({
            "Date":date,
            "Time":time,
            "Sender":Sender,
            "Message":message})
print("parsed messages:",len(messages))


parsed messages: 3174


In [ ]:
# Now let's check if this works correctly
print(messages[-5:])

[{'Date': '30/05/24', 'Time': '19:14', 'Sender': 'Priya', 'Message': 'Take care everyone\n'}, {'Date': '30/05/24', 'Time': '19:28', 'Sender': 'Priya', 'Message': 'Karan that sounds tough, take care\n'}, {'Date': '30/05/24', 'Time': '21:17', 'Sender': 'Aman', 'Message': 'the existential dread is back\n'}, {'Date': '30/05/24', 'Time': '21:30', 'Sender': 'Karan', 'Message': 'Long day guys, woke up at six for that placement workshop which started at eight by the way classic, then ran around campus collecting signatures for the project document, ate lunch at four PM standing in the corridor, came back to hostel and realized I forgot my charger in the lab.\n'}, {'Date': '30/05/24', 'Time': '23:31', 'Sender': 'Aman', 'Message': 'anyone awake?'}]


Feature 2: Group Overview

In [ ]:
#FEATURE.2- GROUP OVERVIEW
participants=set()
for msg in messages:
  participants.add(msg["Sender"])
print("Total participants:",len(participants))
print("participants:",participants)

total_messages = len(messages)
print("Total messages:", total_messages)

total_words=0
for msg in messages:
  total_words+=len(msg["Message"].split())
print("Total words:",total_words)

average_words_per_message=total_words/total_messages
print("Average words per message:",average_words_per_message)

from datetime import datetime
dates = []

for msg in messages:
    date = datetime.strptime(msg["Date"], "%d/%m/%y")
    dates.append(date)

start_date = min(dates)
end_date = max(dates)

print("Start date:", start_date.strftime("%d/%m/%Y"))
print("End date:", end_date.strftime("%d/%m/%Y"))

Total participants: 6
participants: {'Karan', 'Aman', 'Priya', 'Rahul', 'Neha', 'Vikas'}
Total messages: 3174
Total words: 31551
Average words per message: 9.940453686200378
Start date: 01/04/2024
End date: 30/05/2024


In [ ]:
message_count = {}

for msg in messages:
    Sender = msg["Sender"]

    if Sender not in message_count:
        message_count[Sender] = 1
    else:
        message_count[Sender] += 1

print(message_count)

print("\nMessages per participant:")
for Sender in message_count:
    print(Sender, ":", message_count[Sender])

print("\nMessage contribution:")

for Sender in message_count:
    percentage = (message_count[Sender] / total_messages) * 100
    print(Sender, ":", round(percentage, 2), "%")

{'Rahul': 953, 'Priya': 718, 'Karan': 354, 'Neha': 635, 'Aman': 490, 'Vikas': 24}

Messages per participant:
Rahul : 953
Priya : 718
Karan : 354
Neha : 635
Aman : 490
Vikas : 24

Message contribution:
Rahul : 30.03 %
Priya : 22.62 %
Karan : 11.15 %
Neha : 20.01 %
Aman : 15.44 %
Vikas : 0.76 %


In [ ]:
print("=" * 45)
print("           GROUP OVERVIEW")
print("=" * 45)

print("Total Participants :", len(participants))
print("Total Messages     :", total_messages)
print("Total Words        :", total_words)
print("Average Words/Msg  :", round(total_words / total_messages, 2))

print("Date Range         :", start_date.strftime("%d/%m/%Y"),
      "to", end_date.strftime("%d/%m/%Y"))

print("\nMessages per Participant")
print("-" * 30)

for Sender in message_count:
    percentage = (message_count[Sender] / total_messages) * 100
    print(Sender, ":", message_count[Sender],
          "(", round(percentage, 2), "%)")

           GROUP OVERVIEW
Total Participants : 6
Total Messages     : 3174
Total Words        : 31551
Average Words/Msg  : 9.94
Date Range         : 01/04/2024 to 30/05/2024

Messages per Participant
------------------------------
Rahul : 953 ( 30.03 %)
Priya : 718 ( 22.62 %)
Karan : 354 ( 11.15 %)
Neha : 635 ( 20.01 %)
Aman : 490 ( 15.44 %)
Vikas : 24 ( 0.76 %)


Feature 3:Most active day and hour

In [ ]:
from datetime import datetime
date_time = datetime.strptime(
    messages[0]["Date"] + " " + messages[0]["Time"],
    "%d/%m/%y %H:%M"
)

print(date_time)
print("Day:", date_time.strftime("%A"))
print("Hour:", date_time.hour)

2024-04-01 01:17:00
Day: Monday
Hour: 1


In [ ]:
day_count={}
hour_count={}
for msg in messages:
  day=datetime.strptime(msg["Date"],"%d/%m/%y").strftime("%A")
  if day not in day_count:
    day_count[day]=1
  else:
    day_count[day]+=1
  hour=int(datetime.strptime(msg["Time"],"%H:%M").hour)
  if hour not in hour_count:
    hour_count[hour]=1
  else:
    hour_count[hour]+=1
print("Most active day:",max(day_count,key=day_count.get))
print("Most active hour:",max(hour_count,key=hour_count.get))

Most active day: Wednesday
Most active hour: 18


In [ ]:
print("=" * 40)
print("       MOST ACTIVE DAY & HOUR")
print("=" * 40)

print("Most Active Day  : Wednesday")
print("Most Active Hour : 18:00")

       MOST ACTIVE DAY & HOUR
Most Active Day  : Wednesday
Most Active Hour : 18:00


Feature 4:Numpy Activity Heatmap

In [ ]:
import numpy as np

In [ ]:
activity_matrix=np.zeros((7,24),dtype=int)
print("Matrix shape:", activity_matrix.shape)
day_index = {
    "Monday": 0,
    "Tuesday": 1,
    "Wednesday": 2,
    "Thursday": 3,
    "Friday": 4,
    "Saturday": 5,
    "Sunday": 6
}

for msg in messages:
    date_time = datetime.strptime(
        msg["Date"] + " " + msg["Time"],
        "%d/%m/%y %H:%M"
    )

    day = date_time.strftime("%A")
    hour = date_time.hour

    activity_matrix[day_index[day]][hour] += 1

print(activity_matrix)

Matrix shape: (7, 24)
[[ 8 17 13 11 11  0  4 10 18 12 18 15 31 30 18 20 30 23 49 35 30 32 19 28]
 [11 16 13 25 14 12  4  9 18 21 19  7 21 18 20  8 27 26 40 16 21 29 15 16]
 [ 6 11 12  6 11  4  0  8 20 31 25 15 27 17 37 27 38 25 21 34 26 46 19 17]
 [ 7  5 15 14  9  6  1  9 16 23 20 30 37 24 26 18 33 19 36 35 32 18 20 22]
 [ 5 11 17  4 22  3  6  4 21 26 23 11 23 25 18 22 17 17 41 25 15 18 12 17]
 [11 15  9 10 17  0  3  7 14 21 23 14 20 22 25  9 25 40 34 43 20 17 15 45]
 [ 9  7  4  7 26  4 15  8 15 17 32 22 34 23 18 27 19 23 27 40 22 17 16 14]]


In [ ]:
print("Total messages in matrix:", np.sum(activity_matrix))
print("Total messages parsed  :", len(messages))

Total messages in matrix: 3174
Total messages parsed  : 3174


In [ ]:
days = ["Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun"]

print("Activity Heatmap")
print("Day/Hour", end=" ")

for hour in range(24):
    print(f"{hour:2}", end=" ")

print()

for i in range(7):
    print(f"{days[i]:>7}", end=" ")

    for hour in range(24):
        value = activity_matrix[i][hour]

        if value >= 40:
            symbol = "█"
        elif value >= 30:
            symbol = "▓"
        elif value >= 20:
            symbol = "▒"
        elif value >= 10:
            symbol = "░"
        else:
            symbol = "."

        print(f"{symbol:>2}", end=" ")

    print()

Activity Heatmap
Day/Hour  0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 
    Mon  .  ░  ░  ░  ░  .  .  ░  ░  ░  ░  ░  ▓  ▓  ░  ▒  ▓  ▒  █  ▓  ▓  ▓  ░  ▒ 
    Tue  ░  ░  ░  ▒  ░  ░  .  .  ░  ▒  ░  .  ▒  ░  ▒  .  ▒  ▒  █  ░  ▒  ▒  ░  ░ 
    Wed  .  ░  ░  .  ░  .  .  .  ▒  ▓  ▒  ░  ▒  ░  ▓  ▒  ▓  ▒  ▒  ▓  ▒  █  ░  ░ 
    Thu  .  .  ░  ░  .  .  .  .  ░  ▒  ▒  ▓  ▓  ▒  ▒  ░  ▓  ░  ▓  ▓  ▓  ░  ▒  ▒ 
    Fri  .  ░  ░  .  ▒  .  .  .  ▒  ▒  ▒  ░  ▒  ▒  ░  ▒  ░  ░  █  ▒  ░  ░  ░  ░ 
    Sat  ░  ░  .  ░  ░  .  .  .  ░  ▒  ▒  ░  ▒  ▒  ▒  .  ▒  █  ▓  █  ▒  ░  ░  █ 
    Sun  .  .  .  .  ▒  .  ░  .  ░  ░  ▓  ▒  ▓  ▒  ░  ▒  ░  ▒  ▒  █  ▒  ░  ░  ░ 


Feature 5: Word Frequency

In [ ]:
word_frequency={}
for msg in messages:
  words=msg["Message"].split()
  for word in words:
    word=word.lower()
    if word not in word_frequency:
      word_frequency[word]=1
    else:
      word_frequency[word]+=1
print("Unique Words:",len(word_frequency))

Unique Words: 626


In [ ]:
top_words = sorted(
    word_frequency.items(),
    key=lambda item: item[1],
    reverse=True
)

print("Top 10 Words")
print("-" * 30)

for word, count in top_words[:10]:
    print(word, ":", count)

Top 10 Words
------------------------------
the : 1494
and : 1013
i : 759
to : 667
is : 559
was : 400
in : 388
how : 321
this : 313
you : 312


In [ ]:
stop_words = {
    "the", "and", "i", "to", "is", "was", "in", "how", "this","you", "a", "of", "for", "it", "on", "that", "my", "me","we", "are", "be", "have",
    "has", "he", "she", "they","with", "but", "or", "so", "do", "not", "what", "your","from", "at", "as", "an", "if", "can", "just", "all","about",
    "am", "his", "which", "everyone", "today","telling", "guys"
}

meaningful_words = {}

for word, count in word_frequency.items():
    if word not in stop_words:
        meaningful_words[word] = count

top_meaningful_words = sorted(
    meaningful_words.items(),
    key=lambda item: item[1],
    reverse=True
)

print("Top 10 Meaningful Words")
print("-" * 35)

for word, count in top_meaningful_words[:10]:
    print(word, ":", count)

Top 10 Meaningful Words
-----------------------------------
hai : 249
bhai : 160
one : 157
had : 151
started : 150
no : 146
scene : 145
entire : 145
please : 141
yaar : 139


Feature 6: Response Time & Silent Streaks

In [ ]:
first_time = datetime.strptime(
    messages[0]["Date"] + " " + messages[0]["Time"],
    "%d/%m/%y %H:%M"
)

second_time = datetime.strptime(
    messages[1]["Date"] + " " + messages[1]["Time"],
    "%d/%m/%y %H:%M"
)

gap = second_time - first_time

print("First message :", first_time)
print("Second message:", second_time)
print("Time gap     :", gap)

First message : 2024-04-01 01:17:00
Second message: 2024-04-01 01:17:00
Time gap     : 0:00:00


In [ ]:
time_gaps = []

for i in range(1, len(messages)):
    previous_time = datetime.strptime(
        messages[i-1]["Date"] + " " + messages[i-1]["Time"],
        "%d/%m/%y %H:%M"
    )

    current_time = datetime.strptime(
        messages[i]["Date"] + " " + messages[i]["Time"],
        "%d/%m/%y %H:%M"
    )

    gap = current_time - previous_time
    time_gaps.append(gap.total_seconds())

print("Total time gaps:", len(time_gaps))
print("First 10 gaps:", time_gaps[:10])
print("Average of the timegaps:", sum(time_gaps) / len(time_gaps))
print("Longest silent streak:", max(time_gaps))
print("Fastest reesponse gap:", min(time_gaps))

Total time gaps: 3173
First 10 gaps: [0.0, 60.0, 3300.0, 0.0, 60.0, 19920.0, 5640.0, 4920.0, 3300.0, 60.0]
Average of the timegaps: 1631.78064922786
Longest silent streak: 21540.0
Fastest reesponse gap: 0.0


In [ ]:
response_times = []

for i in range(1, len(messages)):

    if messages[i]["Sender"] != messages[i-1]["Sender"]:

        previous_time = datetime.strptime(
            messages[i-1]["Date"] + " " + messages[i-1]["Time"],
            "%d/%m/%y %H:%M"
        )

        current_time = datetime.strptime(
            messages[i]["Date"] + " " + messages[i]["Time"],
            "%d/%m/%y %H:%M"
        )

        gap = (current_time - previous_time).total_seconds()

        response_times.append(gap)

print("Total response gaps:", len(response_times))
print("Average response time:",
      round(np.mean(response_times), 2), "seconds")

print("Fastest response:",
      round(np.min(response_times), 2), "seconds")

print("Longest response:",
      round(np.max(response_times), 2), "seconds")

longest_gap_seconds = np.max(time_gaps)

longest_gap_hours = longest_gap_seconds / 3600

print("Longest silent streak:",
      round(longest_gap_hours, 2), "hours")

Total response gaps: 1380
Average response time: 2469.78 seconds
Fastest response: 0.0 seconds
Longest response: 21540.0 seconds
Longest silent streak: 5.98 hours


Feature 7: Personality Archetype Detection

In [ ]:
from datetime import datetime
import numpy as np

# Calculate Message count for each participant

participant_message_count = {}

for msg in messages:
    Sender = msg["Sender"]

    if Sender not in participant_message_count:
        participant_message_count[Sender] = 1
    else:
        participant_message_count[Sender] += 1

#Calculate total words and average words/Message

participant_word_count = {}

for msg in messages:
    Sender = msg["Sender"]
    words = msg["Message"].split()

    if Sender not in participant_word_count:
        participant_word_count[Sender] = len(words)
    else:
        participant_word_count[Sender] += len(words)


participant_avg_words = {}

for Sender in participant_message_count:
    participant_avg_words[Sender] = (
        participant_word_count[Sender]
        / participant_message_count[Sender]
    )

# Calculate response times for each participant

participant_response_times = {}

for Sender in participants:
    participant_response_times[Sender] = []


for i in range(1, len(messages)):

    previous_Sender = messages[i - 1]["Sender"]
    current_Sender = messages[i]["Sender"]

    # Only consider a change of Sender as a response
    if previous_Sender != current_Sender:

        previous_time = datetime.strptime(
            messages[i - 1]["Date"] + " " + messages[i - 1]["Time"],
            "%d/%m/%y %H:%M"
        )

        current_time = datetime.strptime(
            messages[i]["Date"] + " " + messages[i]["Time"],
            "%d/%m/%y %H:%M"
        )

        response_time = (
            current_time - previous_time
        ).total_seconds()

        participant_response_times[current_Sender].append(
            response_time
        )

#Calculate average response time

participant_avg_response = {}

for Sender in participants:

    response_list = participant_response_times[Sender]

    if len(response_list) > 0:
        participant_avg_response[Sender] = np.mean(response_list)
    else:
        participant_avg_response[Sender] = 0

# Calculate each participant's Message percentage

participant_percentage = {}

for Sender in participant_message_count:

    participant_percentage[Sender] = (
        participant_message_count[Sender]
        / total_messages
    ) * 100

#Find each participant's most-used meaningful word

participant_word_frequency = {}

for Sender in participants:
    participant_word_frequency[Sender] = {}


for msg in messages:

    Sender = msg["Sender"]

    words = msg["Message"].lower().split()

    for word in words:

        word = word.strip(".,!?;:\"'()[]{}")

        if word == "":
            continue

        if word in stop_words:
            continue

        if word not in participant_word_frequency[Sender]:
            participant_word_frequency[Sender][word] = 1
        else:
            participant_word_frequency[Sender][word] += 1


participant_top_word = {}

for Sender in participants:

    word_data = participant_word_frequency[Sender]

    if len(word_data) == 0:
        participant_top_word[Sender] = "N/A"
    else:

        sorted_words = sorted(
            word_data.items(),
            key=lambda item: item[1],
            reverse=True
        )

        participant_top_word[Sender] = sorted_words[0][0]

# Calculate group averages

average_messages_per_person = (
    total_messages / len(participants)
)

average_words_per_Message = (
    total_words / total_messages
)


response_values = []

for Sender in participants:

    if participant_avg_response[Sender] > 0:
        response_values.append(
            participant_avg_response[Sender]
        )


if len(response_values) > 0:
    group_average_response = np.mean(response_values)
else:
    group_average_response = 0

# Assign personality archetypes

personality = {}

for Sender in participants:

    message_count = participant_message_count[Sender]

    avg_words = participant_avg_words[Sender]

    avg_response = participant_avg_response[Sender]

    message_percentage = participant_percentage[Sender]


    # Very active participant
    if message_percentage >= 25:
        archetype = "The Main Character"


    # High message count + short Messages
    elif (
        message_count >= average_messages_per_person
        and avg_words < average_words_per_message
    ):
        archetype = "The Chatterbox"


    # Long Messages
    elif avg_words >= average_words_per_message * 1.5:
        archetype = "The Storyteller"


    # Fast responder
    elif (
        avg_response > 0
        and avg_response <= group_average_response * 0.5
    ):
        archetype = "The Quick Responder"


    # Low activity
    elif message_count < average_messages_per_person * 0.5:
        archetype = "The Silent Observer"


    # Default category
    else:
        archetype = "The Balanced Member"


    personality[Sender] = archetype

# Display final personality analysis

print("=" * 75)
print("              PERSONALITY ARCHETYPE ANALYSIS")
print("=" * 75)

print()

for Sender in sorted(
    participants,
    key=lambda name: participant_message_count[name],
    reverse=True
):

    print("Participant :", Sender)
    print("Messages    :", participant_message_count[Sender])
    print("Contribution:", round(participant_percentage[Sender], 2), "%")
    print("Avg Words   :", round(participant_avg_words[Sender], 2))

    if participant_avg_response[Sender] > 0:
        print(
            "Avg Response:",
            round(participant_avg_response[Sender] / 60, 2),
            "minutes"
        )
    else:
        print("Avg Response: N/A")

    print("Top Word    :", participant_top_word[Sender])
    print("Archetype   :", personality[Sender])

    print("-" * 75)

#Summary

print()
print("=" * 75)
print("                  PERSONALITY SUMMARY")
print("=" * 75)

for Sender in sorted(
    participants,
    key=lambda name: participant_message_count[name],
    reverse=True
):

    print(
        Sender,
        "->",
        personality[Sender]
    )

              PERSONALITY ARCHETYPE ANALYSIS

Participant : Rahul
Messages    : 953
Contribution: 30.03 %
Avg Words   : 2.56
Avg Response: 34.95 minutes
Top Word    : hai
Archetype   : The Main Character
---------------------------------------------------------------------------
Participant : Priya
Messages    : 718
Contribution: 22.62 %
Avg Words   : 4.98
Avg Response: 41.99 minutes
Top Word    : please
Archetype   : The Chatterbox
---------------------------------------------------------------------------
Participant : Neha
Messages    : 635
Contribution: 20.01 %
Avg Words   : 5.27
Avg Response: 39.45 minutes
Top Word    : cant
Archetype   : The Chatterbox
---------------------------------------------------------------------------
Participant : Aman
Messages    : 490
Contribution: 15.44 %
Avg Words   : 4.99
Avg Response: 55.36 minutes
Top Word    : sleep
Archetype   : The Balanced Member
---------------------------------------------------------------------------
Participant : Karan
M

THE FINAL REPORT

In [ ]:
print("=" * 60)
print("              GROUPDNA – WHATSAPP GROUP ANALYSIS")
print("=" * 60)

print("\n1. GROUP OVERVIEW")
print("-" * 60)

print("Group Name          : Hostel Bois 4ever")
print("Total Participants  :", len(participants))
print("Total Messages      :", total_messages)
print("Total Words         :", total_words)
print("Average Words/Msg   :", round(total_words / total_messages, 2))

print("Date Range          :", start_date.strftime("%d/%m/%Y"),
      "to", end_date.strftime("%d/%m/%Y"))

print("\nParticipants:")
for Sender in message_count:
    percentage = (message_count[Sender] / total_messages) * 100
    print("  ", Sender, ":", message_count[Sender],
          "messages (", round(percentage, 2), "%)")


print("\n2. MOST ACTIVE DAY & HOUR")
print("-" * 60)

day_totals = np.sum(activity_matrix, axis=1)

day_names = [
    "Monday", "Tuesday", "Wednesday",
    "Thursday", "Friday", "Saturday", "Sunday"
]

most_active_day_index = np.argmax(day_totals)
most_active_day = day_names[most_active_day_index]

hour_totals = np.sum(activity_matrix, axis=0)
most_active_hour = np.argmax(hour_totals)

print("Most Active Day    :", most_active_day)
print("Most Active Hour   :", str(most_active_hour) + ":00")
print("Messages on Day    :", day_totals[most_active_day_index])
print("Messages at Hour   :", hour_totals[most_active_hour])


print("\n3. TOP WORDS")
print("-" * 60)

sorted_words = sorted(
    word_frequency.items(),
    key=lambda item: item[1],
    reverse=True
)

count = 0

for word, frequency in sorted_words:
    if word not in stop_words:
        print("  ", word, ":", frequency)
        count += 1

        if count == 10:
            break


print("\n4. ACTIVITY SUMMARY")
print("-" * 60)

print("Peak activity occurs on", most_active_day,
      "at", str(most_active_hour) + ":00.")

print("The activity matrix contains",
      np.sum(activity_matrix),
      "messages.")

print("This confirms that all parsed messages")
print("were included in the activity analysis.")


print("\n5. RESPONSE TIME ANALYSIS")
print("-" * 60)

if len(time_gaps) > 0:

    average_gap = np.mean(time_gaps)
    maximum_gap = np.max(time_gaps)

    print("Average Response Gap :", round(average_gap / 60, 2), "minutes")
    print("Longest Silent Gap   :", round(maximum_gap / 3600, 2), "hours")

else:
    print("Response time data unavailable.")


print("\n6. PERSONALITY ARCHETYPES")
print("-" * 60)

print("Personality archetypes are assigned based on")
print("message activity and communication patterns.")

for sender in message_count:

    count = message_count[sender]

    if count >= 700:
        archetype = "The Most Active Member"
    elif count >= 400:
        archetype = "The Regular Contributor"
    elif count >= 100:
        archetype = "The Active Participant"
    else:
        archetype = "The Silent Observer"

    print("  ", sender, ":", archetype)


print("\n" + "=" * 60)
print("                    FINAL SUMMARY")
print("=" * 60)

print("GroupDNA analyzed the WhatsApp group using")
print("Python fundamentals, NumPy and datetime.")

print("\nKey Findings:")
print("- Total messages analyzed :", total_messages)
print("- Total participants      :", len(participants))
print("- Most active day         :", most_active_day)
print("- Most active hour        :", str(most_active_hour) + ":00")
print("- Total words             :", total_words)
print("- Average words/message   :", round(total_words / total_messages, 2))

print("\n" + "=" * 60)
print("                  END OF REPORT")
print("=" * 60)

              GROUPDNA – WHATSAPP GROUP ANALYSIS

1. GROUP OVERVIEW
------------------------------------------------------------
Group Name          : Hostel Bois 4ever
Total Participants  : 6
Total Messages      : 3174
Total Words         : 31551
Average Words/Msg   : 9.94
Date Range          : 01/04/2024 to 30/05/2024

Participants:
   Rahul : 953 messages ( 30.03 %)
   Priya : 718 messages ( 22.62 %)
   Karan : 354 messages ( 11.15 %)
   Neha : 635 messages ( 20.01 %)
   Aman : 490 messages ( 15.44 %)
   Vikas : 24 messages ( 0.76 %)

2. MOST ACTIVE DAY & HOUR
------------------------------------------------------------
Most Active Day    : Wednesday
Most Active Hour   : 18:00
Messages on Day    : 483
Messages at Hour   : 248

3. TOP WORDS
------------------------------------------------------------
   hai : 249
   bhai : 160
   one : 157
   had : 151
   started : 150
   no : 146
   scene : 145
   entire : 145
   please : 141
   yaar : 139

4. ACTIVITY SUMMARY
----------------------

Key Findings and Conclusion

The analysis of the Hostel Bois 4ever WhatsApp group produced the following findings:

- The dataset contains **3,174 messages** from **6 participants**.
- The conversation contains **31,551 words**.
- The average message contains approximately **9.94 words**.
- **Wednesday** is the most active day.
- **18:00** is the most active hour.
- The average gap between consecutive messages is approximately **27.2 minutes**.
- The longest observed silent gap is approximately **5.98 hours**.
- Word-frequency analysis identified frequently used conversational terms such as hai, bhai, one, scene, and yaar.
- Communication archetypes were assigned based on participant activity levels.